### ISSUE TITLE
Promise returned by `FastifyInstance.register()` is resolved before plugin's async registration promise has resolved

### ISSUE THREAD

**Issue by chrskrchr** [fastify/fastify#3726]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

3.27.2

### Plugin version

N/A

### Node.js version

14.18.1

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

12.2.1

### Description

I suspect this may actually be a bug in `avvio`, but logging it here since it's easily reproduced in a Fastify server startup sequence.

We have a handful of Fastify plugins with `async` registration functions.  Our app registers them sequentially via code like:

```
await fastify.register(asyncPlugin1);
await fastify.register(asyncPlugin2);
await fastify.register(asyncPlugin3);
```

We're running into an interesting issue where if we inject some async code in between the plugin register calls (e.g., `await someOtherAsyncFunc()`), the subsequent promises returned by `fastify.register()` appear to be resolved immediately without waiting for internal async registration promises to be resolved.

### Steps to Reproduce

1. Clone this repo: https://github.com/chrskrchr/fastify-plugin-race-condition
2. Run `npm install`
3. Run `node index.js`

Alternatively, run this code yourself:

```
import Fastify from "fastify";

function sleep(timeMs){
  return new Promise((resolve) => setTimeout(resolve, timeMs));
}

(async () => {
  const fastify = Fastify({
    logger: true,
  });

  await fastify.register(async function plugin0() {
    console.log("plugin0 register start");
    console.log("plugin0 register finish");
  });

  await fastify.register(async function plugin1() {
    console.log("plugin1 register start");

    await sleep(500);

    console.log("plugin1 register finish");
  });

  await sleep(1);

  await fastify.register(async function plugin2() {
    console.log("plugin2 register start");

    await sleep(500);

    console.log("plugin2 register finish");
  });

  console.log("starting server");

  await fastify.listen(3000);
})();
```

### Expected Behavior

We'd expect to see the following output where all plugins complete registration prior to the server being started:

```
plugin0 register start
plugin0 register finish
plugin1 register start
plugin1 register finish
plugin2 register start
plugin2 register finish
starting server
{"level":30,"time":1645638489287,"pid":13851,"hostname":"AUS-CKARCHER","msg":"Server listening at http://127.0.0.1:3000"}
```

Instead, we see the following output where the server is started prior to `plugin2` actually completing its async registration:

```
plugin0 register start
plugin0 register finish
plugin1 register start
plugin1 register finish
plugin2 register start
starting server
plugin2 register finish
{"level":30,"time":1645638594807,"pid":13905,"hostname":"AUS-CKARCHER","msg":"Server listening at http://127.0.0.1:3000"}
```

**Comment by climba03003** [comment]
I think it is the problem of `console.log` as it is a async task for writing stdout.
If you should it inside an array then output as once. The sequence is in correct order.

```js
const Fastify = require("fastify");
const { promisify } = require("util");
const sleep = promisify(setTimeout);

(async () => {
  const fastify = Fastify({
    logger: true,
  });
  const arr = [];

  await fastify.register(async function plugin0() {
    arr.push("plugin0 register start");
    arr.push("plugin0 register finish");
  });

  await fastify.register(async function plugin1() {
    arr.push("plugin1 register start");

    await sleep(500);

    arr.push("plugin1 register finish");
  });

  await sleep(1);

  await fastify.register(async function plugin2() {
    arr.push("plugin2 register start");

    await sleep(500);

    arr.push("plugin2 register finish");
  });

  arr.push("starting server");

  console.log(arr.join("\n"));

  await fastify.listen(3000);
})();
```

**Comment by mcollina** [comment]
Looks like a bug in avvio :(. Fixing this is going to be complex  and I will prefer to focus on shipping Fastify v4 instead. 

I think you can work around this by removing most of the awaits and wrapping in-between awaits in plugins.

Of course, a PR that fixes it would be amazing.

**Comment by chrskrchr** [comment]
> I think you can work around this by removing most of the awaits and wrapping in-between awaits in plugins.

Yep - this is exactly how we're working around it in our apps.

> a PR that fixes it would be amazing.

I've started poking through the output produced by `DEBUG=avvio node index.js` and will see if anything obvious jumps out.

**Comment by chrskrchr** [comment]
>  I will prefer to focus on shipping Fastify v4 instead.

@mcollina - do you think this bug will no longer exist in Fastify v4?  Or are you just saying you'd prefer to prioritize getting v4 out the door over fixing this bug that has a handful of workarounds? (which I'd completely understand)

I'm asking because I have a WIP `avvio` branch that appears to fix the issue, but it's somewhat hacky and I don't know if it's something we'd want to live on in future versions of Fastify.  I'd be much more comfortable submitting it for review if it's something I knew would die with Fastify v3.  😄 

I can submit the WIP PR for discussion regardless.  Just want to see if the team would even be willing to consider a somewhat-hacky fix.

(and if you're wondering why I'm even considering a short term fix vs. adopting one of the workaround, it's because we have 30+ fastify apps at my company and it would be much easier and future proof IMO to ask them to bump to a newer version of Fastify w/ a fix rather than asking them all to refactor their Fastify startup sequence)

**Comment by mcollina** [comment]
> @mcollina - do you think this bug will no longer exist in Fastify v4? Or are you just saying you'd prefer to prioritize getting v4 out the door over fixing this bug that has a handful of workarounds? (which I'd completely understand)

This bug exists in Fastify v4. I'm only focusing on Fastify v4 launch now.

> I'm asking because I have a WIP avvio branch that appears to fix the issue, but it's somewhat hacky and I don't know if it's something we'd want to live on in future versions of Fastify. I'd be much more comfortable submitting it for review if it's something I knew would die with Fastify v3. 😄

Go ahead and open that PR. I know it's a hacky fix - the whole `await` for plugin is very hacky and well hidden.

**Comment by chrskrchr** [comment]
PR submitted here: https://github.com/fastify/avvio/pull/173

The "hack" I referred to in my previous comment was actually addressing an issue that's subtly different from the one described in this issue.  It appears that different race conditions can occur depending on whether `autostart` is `true` vs. `false`, and the PR I ultimately submitted above deals with the `autostart: false` case, which is what Fastify is doing internally.

**Comment by chrskrchr** [comment]
I pulled `avvio@7.2.3` into some of our Fastify applications and verified that the fix addresses our issue.  Marking this issue as closed.  🙌

**Comment by mcollina** [comment]
Unfortunately https://github.com/fastify/avvio/pull/173 broke Fastify, I'll revert the fix on both lines.
See https://github.com/fastify/fastify-cors/issues/189#issuecomment-1050950864 for more details.

**Comment by chrskrchr** [comment]
Dang! :(

I'll take a closer look at the `fastify-cors` issue and see if I can find a way to fix.

**Comment by chrskrchr** [comment]
Good news, bad news.

Bad news first - @mcollina was right that [the change in behavior observed in this unit test](https://github.com/fastify/avvio/pull/173#discussion_r814204143) was a symptom of a change (a bug, really) in how avvio was queueing and executing plugin callbacks.  It appears to be a symptom of [the same underlying bug that broke route registration](https://github.com/fastify/fastify-cors/issues/189). 😞 

Good news - I think I have an alternative fix that 1) addresses the original issue described in this thread and 2) appears to not break anything.  I've tested it in both [my test repo](https://github.com/chrskrchr/fastify-plugin-race-condition) and [pvanagtmaal's test repo](https://github.com/pvanagtmaal/fastify-test) and everything looks good.  But I'm going to hold off on submitting a PR until next week so I can test a few more things and make sure I'm not wasting anyone's time.

**Comment by chrskrchr** [comment]
Here's a second attempt at fixing this issue:

https://github.com/fastify/avvio/pull/177

**Comment by chrskrchr** [comment]
The fix above was released in `avvio@7.2.5` and `avvio@8.1.3`.  Marking this as closed.  (and hoping we don't have to re-open again 🤞 😬 )
